modules/ggl-http/src/gghttp_util.c

```diff
@@ -47,6 +54,169 @@ static GglError translate_curl_code(CURLcode code) {
     }
 }
 
+static bool can_retry(CURLcode code, CurlData *data) {
+    switch (code) {
+    // If OK, then inspect HTTP status code.
+    case CURLE_OK:
+        break;
+
+    case CURLE_OPERATION_TIMEDOUT:
+    case CURLE_COULDNT_CONNECT:
+    case CURLE_SSL_CONNECT_ERROR:
+    case CURLE_GOT_NOTHING:
+    case CURLE_SEND_ERROR:
+    case CURLE_RECV_ERROR:
+    case CURLE_PARTIAL_FILE:
+    case CURLE_AGAIN:
+        return true;
+
+    default:
+        return false;
+    }
+
+    long http_status_code = 0;
+    curl_easy_getinfo(data->curl, CURLINFO_HTTP_CODE, &http_status_code);
+
+    switch (http_status_code) {
+    case 400: // Generic client error
+    case 408: // Request timeout
+    // TODO: 429 can contain a retry-after header.
+    // This should be used as the backoff.
+    case 429: // Too many requests
+    case 500: // Generic server error
+    case 502: // Bad gateway
+    case 503: // Service unavailable
+    case 504: // Gateway Timeout
+    case 509: // Server bandwidth exceeded
+        return true;
+    default:
+        return false;
+    }
+}
+
+typedef struct CurlRequestRetryCtx {
+    CurlData *curl_data;
+
+    // reset response_data for next attempt
+    GglError (*retry_fn)(void *);
+    void *response_data;
+
+    // Needed to propagate errors when retrying is impossible.
+    GglError err;
+} CurlRequestRetryCtx;
+
+static GglError retry_reset_buffer(void *response_data) {
+    GglByteVec *vector = (GglByteVec *) response_data;
+    vector->buf.len = 0;
+    return GGL_ERR_OK;
+}
+
+static GglError retry_reset_fd(void *response_data) {
```

Comment:
This function name is a bit inaccurate; its emptying the file (resetting fd sounds like fseeking to 0).